### PR TITLE
Multi-Product compatability

### PR DESCRIPTION
**Pull request by britkat1980**
Includes divenal's updates and mine to make it work with all GE products.

One outstanding item I need is an ability to dump the "model"/register cache to a dict

### CODE REVIEW

**Comment by divenal** [comment]
> Includes divenal's updates and mine to make it work with all GE products.
> 
> One outstanding item I need is an ability to dump the "model"/register cache to a dict

the register cache _is_ a dict ..?

**Comment by britkat1980** [comment]
perhaps I'm not doing it right, but I can't directly put the inverter object into a dict and see all the items. If I try to iterate through items I get an error that Inverter is not iterable... should I be doing something different?

to be clear is not the register_cache, but the inverter object itself

File: givenergy_modbus/model/inverter.py [hunk 1]
```diff
@@ -33,6 +33,20 @@ class Model(StrEnum):
     def _missing_(cls, value):
         """Pick model from the first digit of the device type code."""
         return cls(value[0])
+    
+    @classmethod
+    def add_regs(cls, value):
+        """Return possible additional registers."""
+        regs={
+            '2': ([],[180,240,300,360]),    #Hybrid
```
Comment (divenal) [inline givenergy_modbus/model/inverter.py]: Hmm.. they're all very different, aren't they. How interesting...

I've been playing with the notion of using python sets to express register sets. See https://github.com/divenal/givenergy-modbus-async/commit/ea41f0c92d6ce16ea7b9bf3bd4656be28b318984  (still work in progress)

Are these all possible registers that we need to probe for even after detecting the model type, or once the model has been established, are all these registers known to be there ?  For LV hybrid inverters, there's certainly a detail that later generations have more registers than older generations, so there may still be a need to probe to see what's available ?

File: givenergy_modbus/model/plant.py [hunk 2]
```diff
@@ -96,9 +102,22 @@ def inverter(self) -> Inverter:
...
-        """Return Battery models for the Plant."""
-        return [
-            Battery(self.register_caches[i + 0x32])
-            for i in range(self.number_batteries)
-        ]
+    def batteries(self): # -> list[Battery]:
+        """Return LV Battery models for the Plant."""
+        if self.isHV:
+            return [
+                HVBattery.from_orm(self.register_caches[i + 0x50])
```
Comment (divenal) [inline givenergy_modbus/model/plant.py]: Regression?  I think from_orm was a pydantic thing.

File: givenergy_modbus/model/inverter.py (hunk 1 above)
Comment (britkat1980) [reply givenergy_modbus/model/inverter.py]: This set here is the additional registers which should be probed for each model type. For most they will always exist, for others they must be tested using detect_plant. This list gives the starting point then detect_plant confirms if they exists (which is based on firmware version in reality)

File: givenergy_modbus_async/client/client.py.bkp [hunk 3]
```diff
@@ -153,6 +153,97 @@ async def watch_plant(
                     reqs, timeout=timeout, retries=retries, return_exceptions=True
                 )
 
+    async def detect_plant(
```
Comment (divenal) [inline givenergy_modbus_async/client/client.py.bkp:225]: I'm currently aiming towards keeping all this knowledge in plant.py rather than the client having to worry about it.
https://github.com/divenal/givenergy-modbus-async/commit/ea41f0c92d6ce16ea7b9bf3bd4656be28b318984  has code to implement refresh_plant() that way. I hope for detect_plant() to do something similar, where it asks the plant to suggest registers that need to be fetched. Probably iteratively, since we need to identify the model first. (I'm assuming we first determine model by getting HR(0..59) from address 0x11)

That change also adds a mechanism which might be useful for GivTCP Application code can extend Plant class and hook method registers_updated(). Then each time a WriteHoldingRegisterResponse arrives (due to a write from any connected client, or the portal), that method will be invoked. This way, GivTCP doesn't have to poll holding registers regularly to discover what has changed - it can be told immediately.

File: givenergy_modbus/model/plant.py (hunk 2 above)
Comment (britkat1980) [reply givenergy_modbus/model/plant.py]: Yep, this has been removed now

**Comment by britkat1980** [comment]
> perhaps I'm not doing it right, but I can't directly put the inverter object into a dict and see all the items. If I try to iterate through items I get an error that Inverter is not iterable... should I be doing something different?
> 
> to be clear is not the register_cache, but the inverter object itself

I've added in a "getall" method in the register to iterate round and create the object

File: givenergy_modbus/model/register.py [hunk 4]
```diff
@@ -21,12 +21,30 @@
...
+        return voltage[option]
+
     @staticmethod
     def uint16(val: int) -> int:
         """Simply return the raw unsigned 16-bit integer register value."""
         if val is not None:
             return int(val)
 
+    @staticmethod
+    def gateway_version(first: int,second: int,third: int,fourth: int,) -> Optional[str]:
```
Comment (divenal) [inline givenergy_modbus/model/register.py]: Can/should this be defined in gateway.py if it's only needed by that ?

File: givenergy_modbus_async/pdu/transparent.py [hunk 5]
```diff
@@ -24,7 +24,7 @@ class TransparentMessage(BasePDU, ABC):
 
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
-        self.slave_address = kwargs.get("slave_address", 0x32)
```
Comment (divenal) [inline givenergy_modbus_async/pdu/transparent.py:27]: I assume this is a deliberate move?  I've never been quite sure which addresses are best.  I'm actually wondering whether there's a case for using 0x31 rather than 0x32 for the plant's cache of inverter registers. That keeps them in a separate dictionary from battery #0 which seems like a good thing.

**Comment by divenal** [comment]
I'm perfectly happy for the answer to be no, but...  is there a case for getting a 1.0 released without this extra support. There are dangers of feature creep, or holding up releases until GE release their products, or releasing with untested code. None of which are good.

What is EMS ?  A search finds it on https://givenergy.co.uk/products/sme-battery-system/  which is some sort of commercial system. I would definitely want to make it crystal clear that we would hold no liability if a commercial system relying on this software broke down or anything.  I'd be be tempted to say that this software is not intended for use on commercial systems, that sort of thing...

Is an EMS like an inverter, or it a manager of inverters ?  Is it similar to a gateway (which manages an AIO inverter in some way ?)

There's still a lot of duplication - seems like there should be some common base classes and inheritance.  I'm wondering if we should have some sort of REGISTER_LUT inheritance, rather than complete copies.   Can we assume all inverter-like devices have an identical block at HR(0), so that we can find out model information ?  Seems wrong to instantiate an Inverter instance so that we can query the model type, without knowing if it's actually an Inverter.  Which is why perhaps there should be some sort of base class of Inverter-like things which does have just enough common low-numered HR registers to query, if you see what I mean.

**Comment by divenal** [comment]
> > perhaps I'm not doing it right, but I can't directly put the inverter object into a dict and see all the items. If I try to iterate through items I get an error that Inverter is not iterable... should I be doing something different?
> > to be clear is not the register_cache, but the inverter object itself
> 
> I've added in a "getall" method in the register to iterate round and create the object

~Hmm - that doesn't sound right.   I have wondered whether it's entirely justified in having lots of aliases of registers so that they can be converted to different formats on the fly. Bad when pydantic was instantiating them all in advance, not so bad now that they are converted on demand.  But if you want to convert them all en-mass into a dictionary, that seems to defeat the purpose again.~

Anyway, we could certainly add an iterator to do that.  Basically, you need to interate through the keys in the REGISTER_LUT, and then you can use get() to convert them.  In my replay script (which replays a packet log from a connection, offline), at the end I display all the values:

```
        # Now iterate over all inverter registers, showing last known state
        inv = plant.inverter
        for reg in Inverter.REGISTER_LUT:
            print(reg, inv.get(reg))
```

Ah - that's probably what you've done anyway.

**Comment by britkat1980** [comment]
> I'm perfectly happy for the answer to be no, but... is there a case for getting a 1.0 released without this extra support. There are dangers of feature creep, or holding up releases until GE release their products, or releasing with untested code. None of which are good.
> 
> What is EMS ? A search finds it on https://givenergy.co.uk/products/sme-battery-system/ which is some sort of commercial system. I would definitely want to make it crystal clear that we would hold no liability if a commercial system relying on this software broke down or anything. I'd be be tempted to say that this software is not intended for use on commercial systems, that sort of thing...
> 
> Is an EMS like an inverter, or it a manager of inverters ? Is it similar to a gateway (which manages an AIO inverter in some way ?)
> 
> There's still a lot of duplication - seems like there should be some common base classes and inheritance. I'm wondering if we should have some sort of REGISTER_LUT inheritance, rather than complete copies. Can we assume all inverter-like devices have an identical block at HR(0), so that we can find out model information ? Seems wrong to instantiate an Inverter instance so that we can query the model type, without knowing if it's actually an Inverter. Which is why perhaps there should be some sort of base class of Inverter-like things which does have just enough common low-numered HR registers to query, if you see what I mean.

Happy to release v1 with only standard inverter support. I'm supporting multiple already in GivTCP, so will use my fork for that, but we can catch-up this repo once we've established the sensible approach.

I agree that common base classes for the "standard" registers makes sense with specific ones for the different ones makes most sense.

File: givenergy_modbus_async/client/commands.py [hunk 6]
```diff
@@ -55,7 +55,95 @@ class RegisterMap:
     BATTERY_DISCHARGE_MIN_POWER_RESERVE = 114
     CHARGE_TARGET_SOC = 116
     REBOOT = 163
+    CHARGE_TARGET_SOC_1 = 242
```
Comment (divenal) [inline givenergy_modbus_async/client/commands.py:61]: I want to avoid adding these duplicate definitions, and instead annotate the writable registers in the REGISTER_LUT, with valid=() markers,

File: givenergy_modbus_async/client/commands.py [hunk 7]
```diff
@@ -147,6 +328,9 @@ def set_inverter_reboot() -> list[TransparentRequest]:
     """Restart the inverter."""
     return [WriteHoldingRegisterRequest(RegisterMap.REBOOT, 100)]
 
+def set_active_power_rate(target: int) -> list[TransparentRequest]:
+    """Set max inverter power rate"""
+    return [WriteHoldingRegisterRequest(RegisterMap.ACTIVE_POWER_RATE, target)]
```
Comment (divenal) [inline givenergy_modbus_async/client/commands.py:464]: My plan was to prefer instead to use write_named_register('active_power_rate', target)
which could also do a validity check on the value automatically.

And then rather than having to enter loads of trivial setter methods, I was playing with having them created on the fly, in a similar way to the new inverter model. See https://github.com/divenal/givenergy-modbus-async/commit/05f543f0139741534c4bb322d6cee4bedc7313c1

However, there's now a complication...

File: givenergy_modbus_async/client/commands.py [hunk 8]
```diff
@@ -201,6 +385,22 @@ def set_battery_soc_reserve(val: int) -> list[TransparentRequest]:
         raise ValueError(f"Minimum SOC / shallow charge ({val}) must be in [4-100]%")
     return [WriteHoldingRegisterRequest(RegisterMap.BATTERY_SOC_RESERVE, val)]
 
+def set_car_charge_boost(val: int) -> list[TransparentRequest]:
+    """Set the minimum level of charge to maintain."""
+    # TODO what are valid values? 4-100?
+    val = int(val)
+    if not 0 <= val <= 22000:
+        raise ValueError(f"Charge Boost power ({val}) must be in [0-22kw]%")
+    return [WriteHoldingRegisterRequest(RegisterMap.CAR_CHARGE_BOOST, val)]
```
Comment (divenal) [inline givenergy_modbus_async/client/commands.py:535]: Problem I now have is that CAR_CHARGE_BOOST appears to be an ems register, not an inverter register. Does the car charger appear as an EMS device ?  I don't know how all these things plug together.  Do they all just appear as different devices on one big modbus, or does the client have to connect specifically to the car charger, rather than the inverter, to be able to issue a CAR_CHARGE_BOOST ?

I'm aware that the car charger can be configured either as a standalone unit, or can be under the control of the inverter. Is this only used when it's configured separately ?

Anyway, I don't like the fact that this command is just sitting here and will send the request on any client connection, whether it's to a car charger or something else entirely.  Okay, it will probably get ignored if that register doesn't exist on the targetted device, but something still seems very wrong.

It seems to me that the commands ought to have some idea of what sort of device is connected, in order to be able to issue valid commands. I guess another case in point is that there seems to be EMS_DISCHARGE_SLOTS and DISCHARGE_SLOTS. So the appropriate request surely depends on what you're connected to ???

Comment (britkat1980) [reply givenergy_modbus_async/client/commands.py:535]: CAR_CHARGE_BOOST is a register in the EMS device itself, which sits between multiple inverters/chargers as a supervisory layer. I suggest that we have a few different classes of device:

1. Inverters (Hybrid/AC/3Ph/AIO)
2. Batteries (HV and LV)
3. EMS
4. Gateway

Then each device type can only control and access the right registers (some of which are shared and some of which are unique). This is currently handled in the REGISTER_LUT, so the inverter class doesn't have the valid items for an EMS_DISCHARGE slot. I assume your new "valid" test in the REGISTER_LUT can handle this

Comment (divenal) [reply givenergy_modbus_async/client/commands.py:535]: The validity test is tied to an individual register definition, so that bit's fine. The bit that's currently missing is knowing which REGISTER_LUT to look in.  At the moment, commands.py is completely unaware of what sort of device we're connected to - it's just a collection of static methods -  so it can't (a) choose which register set to consult, and (b) decide whether it's a good idea to send the command.

Does the car charger come with a built-in EMS, or does it have to be supplied as a separate component ?

I don't know much about the Gateway, other than it's an optional add-on for the AIO, and is the bit that provides backup in case of a power cut.   Does the client connect to the gateway, or to the AIO ?  If the client connects to the gateway, can it also control (access) the AIO (as a modbus device ?) through the same connection, or does it have to connect separately to the AIO ?

There are BCU's and BMU's - is a BCU like an inverter, or like a battery, or like a ems / gateway ?  (Or their own class of thing ?)    (Is the AIO like a BCU, with 4 BMU's, or is it still like an inverter ?  Perhaps a BCU is like an inverter ?)

Perhaps I can find answers to these questions in the GivTCP code ?

Comment (divenal) [reply givenergy_modbus_async/client/commands.py:535]: Looking at the installation guides, Gateway manual suggests chargepoint can be plugged into it. Does the Gateway act as an EMS in this case ?

The chargepoint manual says that, still to come, the chargepoint can be wired into a gen-3 hybrid inverter. Would that be acting as an EMS in that case ?  Or just making it available on its modbus ?

Comment (divenal) [reply givenergy_modbus_async/client/commands.py:535]: At the very least, lookup_writable_register() will need to know the device model type. But rather than just telling it "look for this Inverter register", it would be better to go via the plant instance, to use the model detected (or configured). At that almost certainly means going via the client instance.

So simplest is just to move the set_* commands (back ?) into client.py as instance methods. I'm not sure why the design was such that commands returned a list of pdu requests which you then had to send to client.exec() to action. Seems to make more sense that  client.set_charging_limit(x) just goes ahead and sends the request. Which does then mean it has to become async, and caller has to await it.


On a related note, does it still make sense for plant to offer direct inverter acces (ie plant.inverter) ?  In the new world, we don't know if it's an inverter we're attached to, or somethng else.   Perhaps plant.device which will return an instance of Inverter, Gateway, or whatever else it's atached to.

(Does an EMS offer access to the devices attached to it, or do you have to speak to them separately ?)

Comment (hoggyhoggy) [reply givenergy_modbus_async/client/commands.py:535]: I'm on holiday currently so there's a lot to catch up on but the gateways have their own EMS in them (Energy Management System) 
Ontop of that is the new CEMS unit Giv are developing which takes multiple 'old inverters' and conne to to them via USB and then the CEMS unit has full control not the inverter - so you I terfave to the CEMS unit exclusively which then controls the inverters (if you talk to the inverters directly with a CEMS unit attached then bad things happen)

File: givenergy_modbus_async/pdu/transparent.py (hunk 5 above)
Comment (britkat1980) [reply givenergy_modbus_async/pdu/transparent.py:27]: Yes, and 0x32 will get sent to GE cloud, so using 0x31 stops unnecessary data sent to GE servers.

File: givenergy_modbus_async/client/commands.py (hunk 6 above)
Comment (britkat1980) [reply givenergy_modbus_async/client/commands.py:61]: Agreed, moving this to the valid=() markers, just need to trawl through the LUTs (as per consolidation discussion below)

File: givenergy_modbus_async/client/commands.py (hunk 8 above)
Comment (britkat1980) [reply givenergy_modbus_async/client/commands.py:535]: Worth noting, that while the Gateway's have an EMS, there is a product called an EMS (soon to be replaced by the CEMS) which users will understand as a different product.
For an CEMS, as hoggy notes, it should be the only way to access the inverters at a modbus level. (although it will have a restriction that won't get battery cell level data...).
For a 1:1 Gateway, you can get the "normal" AIO registers at the normal addresses. Not sure about multiple inverters on a single Gateway. That I still need to get my head around (am currently testing someones)

Comment (divenal) [reply givenergy_modbus_async/client/commands.py:535]: > so you I terfave to the CEMS unit exclusively which then controls the inverters (if you talk to the inverters directly with a CEMS unit attached then bad things happen)

I'm guessing it's not easy to talk to the inverters directly anyway, since the USB connection to the CEMS replaces the dongle ?

Comment (hoggyhoggy) [reply givenergy_modbus_async/client/commands.py:535]: Yes and no, on Gen 1 removal of the Dongle does somewhat put an end to direct local access (unless you go in directly via RS485 which is something I've been playing with getting to work) 
On Gen2 onwards the dongle is built in HOWEVER they still have a USB port available - and you can infact put a Gen1 dongle into them and run them from that. 
Same for the AIO gateway you are not forced to use the built in dongle should you wish (although why you wouldn't is unclear other than some sort of failover if your LAN failed or something)

File: givenergy_modbus_async/client/commands.py [hunk 9]
```diff
@@ -332,9 +330,16 @@ def set_active_power_rate(target: int) -> list[TransparentRequest]:
...
-def set_calibrate_battery_soc() -> list[TransparentRequest]:
-    """Set the inverter to recalibrate the battery state of charge estimation."""
-    return [WriteHoldingRegisterRequest(RegisterMap.SOC_FORCE_ADJUST, 1)]
+def set_calibrate_battery_soc(val: int) -> list[TransparentRequest]:
+    """Set the inverter to recalibrate the battery state of charge estimation.
+    0- Stop
+    1- Start
+    3- Charge only
+    """
+    if val in (0,1,3):
```
Comment (divenal) [inline givenergy_modbus_async/client/commands.py:472]: Hmm... okay, current check assumes a range of valid values, whereas this wants a discrete set of valid values.

One bodge to extend that would be if 'valid' is a 2-tuple, it's interpreted as a a range, as now, otherwise it's a list of values.  If you want to allow exactly two discrete values, that can be bodged by repeating one.  eg  (1,5)  allows 1 <= x <= 5   whereas (1,5,5)  allows x in (1,5,5)  (ie  x ==1 or x ==5)

A single allowed value, such as for reboot, could then be done as  (100,)   but it's just as easy to do it as (the interval)  (100, 100)

But if this is really to trigger a battery calibration cycle, is it something we're allowed to offer ?  I though only GivEnergy were allowed to trigger calibrations ?

File: givenergy_modbus/model/__init__.py [hunk 10]
```diff
@@ -60,4 +60,8 @@ def from_repr(cls, start: int | str, end: int | str):
             end = f"{end:04d}"
         end_hour = int(end[:-2])
         end_minute = int(end[-2:])
-        return cls(time(start_hour, start_minute), time(end_hour, end_minute))
+        try:
+            return cls(time(start_hour, start_minute), time(end_hour, end_minute))
+        except:
+            # if there's garbage data return midnight
+            return cls(time(0,0), time(0,0))
```
Comment (divenal) [inline givenergy_modbus/model/__init__.py]: I generally prefer to make a fuss (throw an exception) when given garbage, rather than silenty do something else. Since I'd want to highlight to the user that they've done something stupid.  But if this is also used for converting registers read from the inverter, we do need to be a bit more tolerant.

File: givenergy_modbus_async/client/commands.py (hunk 9 above)
Comment (hoggyhoggy) [reply givenergy_modbus_async/client/commands.py:472]: Calibration is automatic from the inverter firmware, so although there's 7 states (from memory) the only actual options are stop (0) start (1) & charge only (3) 

Charge Only is a new one, this may be the elusive slow calibration sometimes run) 
The rest are just indication of current stage until it gradually rattles through each state before setting the battery capacity. 

As for whether we expose it or not we'll probably need to check with GE, however we do need read visibility of it as things like Predbat monitor for it and relinquish control for the duration as it would screw it up.

Comment (hoggyhoggy) [reply givenergy_modbus_async/client/commands.py:472]: Reboot MUST have some form of protection around it. Anything other than valid value (100?) completely wipes the inverter leaving you with a brick. Bad times all round.

Comment (britkat1980) [reply givenergy_modbus_async/client/commands.py:472]: In discussion with GE, It’s ok to allow battery calibration via GivTCP, only those three states

File: givenergy_modbus/model/__init__.py (hunk 10 above)
Comment (britkat1980) [reply givenergy_modbus/model/__init__.py]: Yeah, as it’s hidden in the register get it throws an exception and causes problems, other option is to return none?

File: givenergy_modbus_async/client/commands.py (hunk 9 above)
Comment (hoggyhoggy) [reply givenergy_modbus_async/client/commands.py:472]: So presumably by definition that means OK to expose on this modbus library? (I think Giv sometimes forgets there's 2 layers to all this)

**Comment by britkat1980** [comment]
Yes I think so

File: givenergy_modbus_async/client/commands.py (hunk 9 above)
Comment (divenal) [reply givenergy_modbus_async/client/commands.py:472]: > Calibration ...
> As for whether we expose it or not we'll probably need to check with GE, however we do need read visibility of it as things like Predbat monitor for it and relinquish control for the duration as it would screw it up.

Yes, I was really asking about whether we wanted to be offering access to set it. Read access should be less of an issue.

File: givenergy_modbus/model/__init__.py (hunk 10 above)
Comment (divenal) [reply givenergy_modbus/model/__init__.py]: > other option is to return none?

My gut feeling would be to prefer that, though of course you have much more experience with actually using this software.

I changed the default for any unavailable register from 0 to None, since it's just too confusing to be told a value is 0 when in fact it's unknown.

File: givenergy_modbus_async/client/commands.py (hunk 9 above)
Comment (divenal) [reply givenergy_modbus_async/client/commands.py:472]: > Reboot MUST have some form of protection around it. Anything other than valid value (100?) completely wipes the inverter leaving you with a brick. Bad times all round.

Even though I'd want to avoid duplication of any regiser definitions, I do wonder if we should make an exception for this one, and have the constructor for WriteHoldingRegisterRequest() absolutely forbid creating a request to write anything other than 100 to that register.

Comment (divenal) [reply givenergy_modbus_async/client/commands.py:472]: > Hmm... okay, current check assumes a range of valid values, whereas this wants a discrete set of valid values.
> 
> One bodge to extend that would be if 'valid' is a 2-tuple, it's interpreted as a a range, as now, otherwise it's a list of values. If you want to allow exactly two discrete values, that can be bodged by repeating one. eg (1,5) allows 1 <= x <= 5 whereas (1,5,5) allows x in (1,5,5) (ie x ==1 or x ==5)
> 

A more general approach (with less risk of cockups) would be for 'valid' to be either a 2-tuple (min, max)  or a callable. Then it can be a lambda with whatever test is required.  so eg this could be

       "soc_force_adjust": Def(C.uint16, BatteryCalibrationStage, HR(29), valid = lambda val : val in (0,1,3 )),

then the lookup can use  callable(valid)  to determine whether 'valid' is a lambda or a simple (min, max) range.

Come to think of it, that could remove the need for the special-casing I currently have which assumes that if the upper limit is 2359 then it must be a time, and therefore needs an additional check that minutes < 60.  That can be done more explicitly using a validity funtion instead.

**Comment by divenal** [comment]
Okay, I think the issue with the register tables can be resolved without too much mess.  To recap, I added commands.write_named_register(name : str, value: int)   which calls Inverter.lookup_named_register()  to map name to number, and verify that the proposed value is valid.  But that hardcodes Inverter, whereas now we want to split that into multiple different device classes depending on what is detected at runtime.

Currently 'commands' is just a module namespace - commands.set_xyz() is just a simple static method living in the 'commands' namespace.

Step 1 is just to recognise that we can change commands to be an instance of some class Commands, and make set_xyz() a member function.  So all the set_xyz() function definitions acquire a first parameter 'self', but no change at the point of call. (As long as the caller has access to a symbol 'commands' which is a singleton instance of Commands.)

Step 2, then, is for the the instance of 'commands' to be tied to the plant instance. In java, this would effectively be an "inner class" - it adds its own commands, but is also a proxy to get back to the plant instance. But this means that now,  commands.write_named_register() can refer to the plant, and get redirected to the correct register lookup table depending on what has been detected. So commands that are not available on the device you're connected to can be rejected, rather than sending a value to a non-existent (or worse, reassigned) register.

As a horrible bodge, while technically client and plant are not singletons,  in practise there's usually only one client and so one plant instance, and so when we instantiate a plant, we could just poke it into the global commands instance, and then commands.plant does still have access to a plant instance. Just by a less pleasing route.
 
TL;DR - it's still fine to go ahead and change the commands.set_xyz() fns to use commands.write_named_register(), and then when we are ready to adopt the multi-product compatibilty, that won't cause very much disruption to existing code.

And https://github.com/divenal/givenergy-modbus-async/commit/05f543f0139741534c4bb322d6cee4bedc7313c1 - my change to magically create commands.set_xyz() from the register lut on the fly - gets closer to being viable.  The automatically-generated docstring bit won't work, though, since, again, that doesn't have a master list of writable registers to consult.  Though I guess it can just list them against each known device separately.

**Comment by divenal** [comment]
The current command model is that you do   client.exec(commands.set_xxx())  which I've always thought as slightly odd... there's not much you can do with the return from commands.set_xxx() other than send it to client.exec().

It occurred to me that a side effect of changing 'commands' to be an object which has access back to the client is that   client.commands.set_xyz()  could actually exec() the requests directly, rather than just creating a list of request that need to be exec-ed.   Well, it would need to become  await client.commands.set_xxx  I suppose.

(It does make newly-resurrected pytest slightly more awkward, but that's just a case of faking up a client with a dummy exec() which checks that the generated commands are correct.)

**Comment by divenal** [comment]
I can't remember if I've asked this... if you have an AIO with Gateway, do these two devices live on different IP addresses, or does one IP address access both devices on different modbus addresses, or does the gateway make the AIO's registers available as part of its own register address space, or something else ?

**Comment by britkat1980** [comment]
So the gateway and AIO are separate devices, each with its own IP and set of Modbus registers. In the scenario where a Gateway is attached to a single AIO then you can collect all the AIO data through the Gateway, which mirrors the “standard” register set. In this case there is little need to connect to the AIO, except to get battery data.
Where the Gateway is in parallel mode, you can collect high level data from the Gateway about each AIO but not the full register set and so you would then connect to the AIOs directly as well…
Also you expose controls for the gateway only in parallel mode as in single mode the gateway does not directly control the AIO

File: givenergy_modbus_async/model/baseinverter.py [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+from .register import (
+    Converter as C,
+    DynamicDoc,
+    HR,
+    IR,
+    RegisterDefinition as Def,
+    RegisterGetter,
+)
+
+class BaseInverter(RegisterGetter, metaclass=DynamicDoc):
```
Comment (divenal) [inline givenergy_modbus_async/model/baseinverter.py:19]: There's a _lot_ of registers defined in this BaseInverter - I assumed the base device class would be a lowest common denominator.  Do all possible devices (including EMS) have things like charge/discharge slots ?  I assumed those would only be for inverter-like devices.

Well, I guess the top-level split is devices you can connect the client to (LV inverter, 3-phase inverter, aio, gateway, ems, ...) and devices that attach to those devices (batteries, bmu, bcu, chargepoint, ...).

And that latter category might split further into those which have their own modbus presence (battery, bmu, bcu), and the chargepoint which exists only as registers in the EMS.  (What happens with multple chargepoints ?)   Not sure if chargepoint is physically wired via modbus, but I get the impression that, even if it is, it's handled internally rather than exposed via tcp interface


Does the EMS have its own grid connection ?  Obviously, it will be powered from AC, but does actively send power out on AC?   Similarly, I know the AIO gateway provides power backup in powercut, but does it actually do so itself, or just switch relays and things for the connected AIO ?  I was just pondering whether a common theme among the devices the client can connect to was how inverter-like they are.

**Comment by divenal** [comment]
My gen-3 hybrid has just been updated to firmware #309. Release notes offer:

1. Add Additional PV total registers, for external PV added to internal PV (Needs Portal update to read)
2. Add Inverter Mode register. Hybrid/String (Supress battery error of not present and intended to be that way)

Do we have details ?

https://kb.givenergy.cloud/article.php?id=92&oid=23

**Comment by hoggyhoggy** [comment]
Hi Both (@britkat1980 & @divenal)
I can see that everyone has been busy and I thought best (before each party goes too far down each path/redoing work already done) that it might be a good idea to get a "latest" GivTCP Dev Build on here to compare notes. Britkat appears to have been very busy, particularly around newer kit register info and so on.  
Copy is in Dev2 Tree currently, with my sole change from the raw docker image being in regards to Unix Linefeed so it picks up changes from this library correctly. 
https://github.com/hoggyhoggy/givenergy-modbus-async/tree/Dev2-(21_07_2024)

**Comment by divenal** [comment]
Yes, I was aware of lots of changes under model, so I've been mostly avoiding that part of the tree, and working on the other bits. Though I had just about reached the point where the next thing to do was to integrate these various inverter flavours.

**Comment by divenal** [comment]
First observation: meter.py introduces two new device classes: Meter and MeterProduct. The former contains only IR's, the latter only MR's. But for other devices we mix IR's and HR's together in a single RegisterGetter class. Is there a particular reason for Meter and MeterDevice being separate?  They both occupy slave addresses 1..8

(I'm also vaguely intrigured that it was necessary to introduce a new register type, rather than sticking with HR. I'm guessing they're read-only, but doesn't see an issue to just have the device reject attempts to write to those registers.)

**Comment by hoggyhoggy** [comment]
> First observation: meter.py introduces two new device classes: Meter and MeterProduct. The former contains only IR's, the latter only MR's. But for other devices we mix IR's and HR's together in a single RegisterGetter class. Is there a particular reason for Meter and MeterDevice being separate? They both occupy slave addresses 1..8
> 
> (I'm also vaguely intrigured that it was necessary to introduce a new register type, rather than sticking with HR. I'm guessing they're read-only, but doesn't see an issue to just have the device reject attempts to write to those registers.)

I'm reading between the lines (and I've yet to cross reference with datasheets for respective manufacturers) but I highly suspect that is a direct interface to the EM115 / GEM120CT products? The registers look familiar so I'm wondering if it is some sort of new idea to get faster reads (perhaps not rely on inverter as much?) and get directly from the meter instead things like grid voltage, import/export and so on. I guess it would be a lot faster and less constrained than polling the inverter? Maybe the class is to try and differentiate it's a Meter (with potentially different brands working differently - EM115, GEM120, whatever the 3 phase one is called...) and not an Inverter / Giv Product?

I may be thinking to far outside the box, it might be something related to the CEMS/EMS magic black box...

**Comment by divenal** [comment]
> Maybe the class is to try and differentiate it's a Meter (with potentially different brands working differently - EM115, GEM120, whatever the 3 phase one is called...) and not an Inverter / Giv Product?

Yes, it's certainly necessary to have a separate subclass of RegisterGetter to distinguish it from an inverter. I'm just not clear why there are separate classes for the IR's and the MR's, whereas inverter-like devices mix IR's and HR's in the same register definition map.
